File: Physical_systems/first lab/Task1/Task1/harmosc_euler.f
      PROGRAM one_d_motion2

      IMPLICIT NONE

      INCLUDE 'globals.inc'

      character(len=32) :: ARG
      integer, parameter :: IN = 1, NCYCLES = 2
      integer :: N = 100, I
      real(db), allocatable :: T(:), X(:), V(:)
      real(db), allocatable :: XANAL(:), VANAL(:)
      real :: DT


!     Number of steps can be given as input argument
      IF (IARGC() > 0) THEN
        CALL GETARG(1,ARG)
        READ(ARG,'(I10)') N
      ENDIF
      ! cycles (of 2*pi)
      DT = NCYCLES*2.0*pi/N
      WRITE(*,*) '#INFO: nr. of steps: ',N
      WRITE(*,*) '#INFO: nr. of cycles: ',NCYCLES
      WRITE(*,*) '#INFO: time step: ',DT
      WRITE(*,*) '#       time            x              v     &
&       x-analytical     v-analytical' 

!     Allocate arrays
      ALLOCATE(T(N+1),X(N+1),V(N+1))
      ALLOCATE(XANAL(N+1),VANAL(N+1))

!     Initial values
      X(1) = 0.0
      V(1) = 1.0
      T(1) = 0.0
      XANAL(1) = X(1)
      VANAL(1) = V(1)

      DO I = 1,N
         T(I+1) = I*DT

!     Analytic solution
         XANAL(I+1) = SIN(T(I+1))
         VANAL(I+1) = COS(T(I+1))

!     Euler (predictor) for position and velocity
         X(I+1) = X(I) + V(I)*DT
         V(I+1) = V(I) - X(I)*DT

      END DO

      WRITE (*,'(5F16.8)') (T(I),X(I),V(I),XANAL(I),VANAL(I),I=1,N+1,IN)

      DEALLOCATE(X,V,T)

      END

File: Physical_systems/first lab/Task1/Task1/harmosc_euler.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>

/* a naive port from harmosc_euler.f
 * Wei Chen, 2013 */

int main(int argc, char *argv[])
{
    int in = 1, ncycles = 2, n = 100, i;
    if (argc == 2) n = atoi(argv[1]);
    double t[n+1], x[n+1], v[n+1], xanaly[n+1], vanaly[n+1];
    double dt, pi;
    /* t[]: time
       x[]: displacement
       v[]: velocity
       xanaly, vanaly: analytical solutions*/

    pi = 4.0*atan(1.0);
    dt = ncycles*2.0*pi/n;   

    printf("#INFO: number of steps:  %d\n", n);
    printf("#INFO: number of cycles: %d\n", ncycles);
    printf("#INFO: time step:        %10.6f\n", dt);
    printf("#       time                x               v           x-analytical      v-analytical\n");

    x[0] = 0.0; v[0] = 1.0; t[0] = 0.0;
    xanaly[0] = x[0]; vanaly[0] = v[0];

    for (i = 1; i < n+1; i++) 
        {
            t[i] = i*dt;
            xanaly[i] = sin(t[i]);
            vanaly[i] = cos(t[i]);
            x[i] = x[i-1] + v[i-1]*dt;
            v[i] = v[i-1] - x[i-1]*dt;
        }

    for (i = 0; i < n+1; i++)
        { 
            printf("%16.8f %16.8f %16.8f %16.8f %16.8f\n", t[i], x[i], v[i], xanaly[i], vanaly[i]);
        }
        
}
